### PR TITLE
FXML-2302: Support for negating PDLL constraints

### PR DESCRIPTION
**Pull request by martin-luecke**
This PR introduces support for negation of constraints.

Previously, this could only be handled by creating an additional native call, e.g.
```PDLL
Constraint checkA(input: Attr);
Constarint checkNotA(input: Attr);
```
or by including an explicit additional operand for negation, e.g.
`Constraint checkA(input: Attr, negated: Attr);`

With this a constraint can simply be negated by prefixing it with `!`. e.g.
```PDLL
Constraint simpleConstraint(op: Op);

Pattern example with benefit(1) {
    let inputOp = op<test.bar>() ->(type: Type);
    let root = op<test.foo>(inputOp.0) -> ();
    !simpleConstraint(inputOp);
    simpleConstraint(root);
    erase root;
}
``` 
This pattern only matches if the constraint `simpleConstraint` is true for `root` and false for `inputOp`.

This required the following changes:
PDLL:
- make lexer aware of `!` tokens
- support parsing ! as prefix for native constraints (only!)
- extend CallExpr with support for `isNegated`

PDL & PDLInterp Dialects
- adding an attribute `isNegated` with default value `false` to the operations `pdl.apply_native_constraint` and `pdl_interp.apply_constraint`
- make PDLToPDLInterpPass aware of this attribute

ByteCode:
- add isNegated to the generated bytecode
- When interpreting `pdl_interp.apply_constraint` branch depending on `isNegated`

This work is to be upstreamed to llvm-project

### CODE REVIEW

File: mlir/include/mlir/Dialect/PDLInterp/IR/PDLInterpOps.td [hunk 1]
```diff
@@ -100,8 +100,7 @@ def PDLInterp_ApplyConstraintOp : PDLInterp_PredicateOp<"apply_constraint"> {
     pdl_interp.apply_constraint "myConstraint"(%input, %attr, %op : !pdl.value, !pdl.attribute, !pdl.operation) -> ^matchDest, ^failureDest
     ```
   }];
-
-  let arguments = (ins StrAttr:$name, Variadic<PDL_AnyType>:$args);
+  let arguments = (ins StrAttr:$name, Variadic<PDL_AnyType>:$args, DefaultValuedAttr<BoolAttr,"false">:$isNegated);
```
Comment (ttjost) [inline mlir/include/mlir/Dialect/PDLInterp/IR/PDLInterpOps.td:103]: nit:  could you make sure to only add the `DefaultValuedAttr` and not change the line?

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 2]
```diff
@@ -1827,6 +1832,11 @@ FailureOr<ast::Expr *> Parser::parseExpr() {
     case Token::dot:
       lhsExpr = parseMemberAccessExpr(*lhsExpr);
       break;
+    // case Token::exclam:
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:1835]: Should this be deleted?

File: mlir/test/mlir-pdll/Parser/expr.pdll [hunk 3]
```diff
@@ -50,6 +50,22 @@ Pattern {
...
+// CHECK: `-PatternDecl {{.*}}
+// CHECK:   -CallExpr {{.*}} Type<Tuple<>> negated
+// CHECK:     `-DeclRefExpr {{.*}} Type<Constraint>
+// CHECK:       `-UserConstraintDecl {{.*}} Name<checkOp> ResultType<Tuple<>>
+Constraint checkOp(op: Op);
+
+Pattern {
+  let inputOp = op<my_dialect.bar>;
+  !checkOp(inputOp);
+  erase inputOp;
```
Comment (ttjost) [inline mlir/test/mlir-pdll/Parser/expr.pdll:64]: What would happen if we tried to use `!` in the rewrite section?

**Comment by martin-luecke** [comment]
Thank you for the review!
I added the test 👍

**Review by ttjost — APPROVED** [review]
Great! Thanks for adding the test and making the necessary changes.

**Review by ljfitz — APPROVED** [review]
Thanks Martin - looks like a useful addition and clearly coded.

**Pull request merged**